Add comprehensive test suite for ERC20Facet and LibERC20

adamgall opened the pull request:
Closes #12 
Closes #59 

This PR adds a complete test suite with 93 tests covering all functionality of ERC20Facet and LibERC20, including edge cases and security considerations.

## Test Coverage

### ERC20Facet Tests (63 tests)
- Metadata: name, symbol, decimals, totalSupply
- Transfers: transfer, transferFrom with all validations
- Approvals: approve, allowance edge cases, unlimited special cases
- Burn: burn, burnFrom with correct balance tracking
- EIP-2612 Permit: signature validation, nonce handling, replay protection
- DOMAIN_SEPARATOR: chain fork detection
- Security: overflow protection, zero address validation
- Fuzz Testing: comprehensive fuzzing of all core functions

### LibERC20 Tests (37 tests)
- Internal Functions: direct testing of library functions
- Minting: mint with overflow protection
- Complete Flows: mint → transfer → burn sequences
- Edge Cases: zero amounts, self-transfers, maximum values

## Architecture

### Test Harnesses
- ERC20FacetHarness: Adds initialize() and mint() for testing the Facet
- LibERC20Harness: Exposes internal library functions as external

### Organization
- Separate test files for Facet (ERC20Facet.t.sol) and Library (LibERC20.t.sol)
- Clean separation of concerns with focused test suites
- Comprehensive documentation in test/README.md

## Key Features
- ✅ All 100 tests passing
- ✅ Full coverage of standard ERC20 functionality
- ✅ Complete EIP-2612 permit implementation tests
- ✅ Extensive edge case and security testing
- ✅ Fuzz testing for critical paths
- ✅ Well-documented testing approach

This implementation provides thorough test coverage and follows best practices for testing diamond facets and their libraries.

mudgen commented on test/README.md:
```diff
@@ -0,0 +1,272 @@
+# Compose Test Suite
```
This documentation looks great.  I'm thinking this documentation should go into the CONTRIBUTING.md file (and what's currently in CONTRIBUTING.md should go in README.md), because we are keeping READM.md with more concise information and more detailed information in CONTRIBUTING.md.

adamgall replied:
however, to give my own opinion:

- README.md
  - imo, shouldn't include "codebase implementation details", like explaining how tests work or what their directory structure is.
  - i would even say that the whole "Quick Start" section should be rewritten to speak to the end user of this library, not the developer building it. Those instructions should move into CONTRIBUTING.md
  - this file should target end users: library consumers. Explain what the project is, who it's for and why it helps them, clear instructions on how to "install it" or "use it", whatever that might look like, some examples of what an implementation into a client codebase looks like, and some links to external docs (when those exist) and CONTRIBUTING.md and legal stuff, etc.
  - I'm taking inspiration from https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/README.md
- CONTRIBUTING.md
  - i like how this is right now. it's very full of information for other developers working on this codebase
  - for as important as testing is, IMO it commands its own dedicated document (test/README/md).
  - the "testing" section in this document calls out the high level structure and important parts, then directs the reader/developer to the specific test/README.md doc for full details on how to follow this project's testing conventions.
- ./test/README.md
  - Keep mostly all of this content (there's some slop at the end that we can probably delete), right here in this file.
  - it's extensive enough that IMO it would clutter CONTRIBUTING.md, if we merged it in.

mudgen commented:
@adamgall I changed the directory structure.  Please update the directory paths to LibERC20.sol etc.

adamgall commented:
> @adamgall I changed the directory structure. Please update the directory paths to LibERC20.sol etc.

Merged and fixed. All tests passing still. Looks like the new slither job is failing though. Seems out of scope for this PR to fix those issues here.

mudgen commented:
Looks great, thanks @adamgall!
